Add configuration flow for Buienradar integration

RobBie1221 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
This PR will deprecate yaml support for Buienradar integration and add UI configuration. The yaml configuration will be imported once. When users have a weather and sensor setup for same coordinates, they will be merged into one ConfigEntry. Only one camera configuration will be imported.

Following things are changed:
- The `forecast` key of the weather platform is deprecated, forecast data is fetched in the same API call and to be in line with ADR-0003, all available data is exposed.
- Weather, sensor and camera domain are setup in one ConfigEntry. For new entries, sensor and camera entities are disabled by default.
- Dimension for camera images is fixed to 700 pixels

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR implements a basic configuration flow for the Buienradar integration. The integration currently includes the weather, camera and sensor platform. With the flow in this PR, you can setup all three in one configuration flow. The user can choose whether to include the radar image as camera, because this is fetched as image with a separate API call. The user is able to configure multiple instances as long as the lattitude/longitude is unique.

Furthermore, a "Name" attribute is now mandatory, which is used to create entities. In the Integrations page, the Buienradar card will show up with this configured name.

Setting options will be addressed in a later PR.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14000

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RobBie1221 committed:
Add configuration flow for Buienradar integration

RobBie1221 commented:
Work in progress, config flow is fully implemented. Todo is change the current tests to work again and add tests for the configuration flow

RobBie1221 committed:
Update buienradar camera tests to work with config flow

RobBie1221 committed:
Update buienradar weather tests to work with config flow

RobBie1221 committed:
Update buienradar sensor tests to work with config flow

RobBie1221 committed:
Remove buienradar config_flow tests to pass tests

RobBie1221 committed:
Add config flow tests for buienradar integration

RobBie1221 committed:
Increase test coverage for buienradar config_flow tests

RobBie1221 commented:
This is basically done. See limitations in description, but first I'd like to get this approved. Also the weather writes to hass.data[buienradar_condition] which I think should be refactored to hass.data[DOMAIN]. The unload will most likely leave the data updater intact, which is not really nice when removing the integration without restart. I'll address this as well in a later PR.

RobBie1221 commented:
@mjj4791 @ties Mind taking a look at this as well?

RobBie1221 committed:
Move data into domain

RobBie1221 committed:
Remove forecast option

RobBie1221 committed:
Move data to options

RobBie1221 committed:
Remove options from config flow

RobBie1221 committed:
Adjust tests

RobBie1221 committed:
Adjust string

RobBie1221 committed:
Fix pylint issues

OnFreund commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+                f"{user_input.get(CONF_LATITUDE)}-{user_input.get(CONF_LONGITUDE)}"
+                not in configured_instances(self.hass)
+            ):
+                name = user_input[CONF_NAME]
+
+                data = {
+                    CONF_NAME: name,
+                    CONF_LATITUDE: user_input[CONF_LATITUDE],
+                    CONF_LONGITUDE: user_input[CONF_LONGITUDE],
+                    CONF_CAMERA: user_input[CONF_CAMERA],
+                    CONF_SENSOR: True,
+                    CONF_WEATHER: True,
+                    CONF_COUNTRY: user_input[CONF_COUNTRY],
+                }
+
+                return self.async_create_entry(title=name, data=data)
```
I would extract this to a separate function so the logic is easier to follow here.

OnFreund commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+
+    def __init__(self):
+        """Init MetFlowHandler."""
+        self._errors = {}
```
Any reason for this to be part of the class and not local to `async_step_user`?

OnFreund commented on homeassistant/components/buienradar/translations/en.json:
```diff
@@ -0,0 +1,21 @@
+{
```
This file should be created automatically. No need to add it

OnFreund commented on tests/components/buienradar/test_config_flow.py:
```diff
@@ -0,0 +1,174 @@
+        CONF_COUNTRY: "NL",
+    }
+
+    await hass.async_block_till_done()
+
+    conf_entries = hass.config_entries.async_entries(DOMAIN)
+    await hass.config_entries.async_unload(conf_entries[0].entry_id)
+    await hass.async_block_till_done()
```
There are no assertions here. Is this code needed?

OnFreund commented on tests/components/buienradar/test_config_flow.py:
```diff
@@ -0,0 +1,174 @@
+        CONF_COUNTRY: "NL",
+    }
+
+    await hass.async_block_till_done()
+
+    conf_entries = hass.config_entries.async_entries(DOMAIN)
+    await hass.config_entries.async_unload(conf_entries[0].entry_id)
+    await hass.async_block_till_done()
```
No assertions here as well

OnFreund commented on tests/components/buienradar/test_config_flow.py:
```diff
@@ -0,0 +1,174 @@
+    assert result["step_id"] == "user"
+    assert result["errors"] == {CONF_NAME: "name_exists"}
+
+    conf_entries = hass.config_entries.async_entries(DOMAIN)
+    await hass.config_entries.async_unload(conf_entries[0].entry_id)
+    await hass.async_block_till_done()
```
same

RobBie1221 committed:
Rework review comments

OnFreund reviewed (approved):
This looks good to me, but keep in mind someone with write access will have to approve as well.

RobBie1221 commented:
Waiting review, not stale

RobBie1221 committed:
Handle import

RobBie1221 committed:
Change config flow to setup camera or weather

RobBie1221 committed:
Fix tests

RobBie1221 committed:
Remove translated file

RobBie1221 committed:
Fix pylint

RobBie1221 committed:
Fix flake8

RobBie1221 committed:
Fix unload

Kars-de-Jong reviewed (changes requested):
Some minor comments about comments ;-)

RobBie1221 commented on homeassistant/components/buienradar/__init__.py:
```diff
@@ -1 +1,135 @@
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up buienradar2 from a config entry."""
```
Ah, I ran the scaffold with buienradar2, forgot to change it.

RobBie1221 committed:
Minor name changes

ties commented:
I think this still looks relevant functionally. @RobBie1221?

RobBie1221 commented:
Yes, just waiting for a review.

FeikoJoosten commented:
@RobBie1221 would it be possible to add the proposed changes from my [PR](https://github.com/FeikoJoosten/home-assistant.io/pull/1) to this PR? 

I don't mind waiting till this PR is merged to do it myself, but since the changes are small I figured I might ask you for it.

RobBie1221 commented:
Frankly speaking, I tried to keep this PR as small as possible and already it's somewhat large, which is probably the reason why nobody bothers to review it. I wouldn't want to add new functionality as that would not speed up the review process.

FeikoJoosten commented:
No problem at all. I'll await the merge and add it myself. Thanks anyways!

ties commented on CODEOWNERS:
```diff
@@ -76,7 +76,7 @@ homeassistant/components/brother/* @bieniu
 homeassistant/components/bsblan/* @liudger
 homeassistant/components/bt_smarthub/* @jxwolstenholme
-homeassistant/components/buienradar/* @mjj4791 @ties
+homeassistant/components/buienradar/* @mjj4791 @ties @Robbie1221
```
:+1:

ties commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+
+
+class BuienradarFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for buienradar2."""
```
```suggestion
    """Handle a config flow for buienradar."""
```

RobBie1221 committed:
Update homeassistant/components/buienradar/config_flow.py

Co-authored-by: Ties de Kock <ties@tiesdekock.nl>

RobBie1221 committed:
Remove asynctest

frenck commented on homeassistant/components/buienradar/sensor.py:
```diff
@@ -227,8 +211,8 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     dev = []
-    for sensor_type in config[CONF_MONITORED_CONDITIONS]:
-        dev.append(BrSensor(sensor_type, config.get(CONF_NAME), coordinates))
+    for sensor_type in SENSOR_TYPES:
+        dev.append(BrSensor(sensor_type, config[CONF_NAME], coordinates))
```
I think we should disable these sensors by default, as I now end up with 107 entities.

![image](https://user-images.githubusercontent.com/195327/109531168-db31a600-7ab7-11eb-9485-c56f2dd49900.png)

Some might be helpful at some point for some people, but they can activate the needed sensors at that point.

frenck commented on homeassistant/components/buienradar/strings.json:
```diff
@@ -0,0 +1,35 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Buienradar",
```
"Buienradar" is a brand name, it doesn't need to be translated.
Applies to other occurrences in this file as well.

frenck reviewed (changes requested):
A couple of initial comments (mainly from a user experience perspective at this point).

RobBie1221 commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_DOMAIN): vol.In(["Camera", "Weather"]),
```
The main reason to separate is that the camera and weather platform poll a completely different API.

In hindsight, I think disabling an entity would be a good alternative for users who e.g. don't want their device to poll a camera endpoint.

RobBie1221 replied:
Maybe one thing, the weather API is useless for Belgian users if I remember correctly.

RobBie1221 replied:
I'm open to address changes requested to a reasonable level of effort, but if the answer is split it into multiple integrations I have to abandon this and leave it in it's current state...

frenck replied:
> but if the answer is split it into multiple integrations I have to abandon this and leave it in it's current state...

That was the suggestion yes. To be honest, that would make the code a lot easier as well.

RobBie1221 replied:
I can follow the reasoning and do not disagree on that, I'm just saying I'm not going to do that. I've updated the existing integration, where everything is in one integration, to get rid of platform setup and yaml. Imho, this PR is not the smallest one to avoid breaking changes, but aside from that the config flow code is overseeable and I think beneficial for new users stepping into Home Assistant who are not familiar with yaml. If that's not accepted in it's current form, than it is what it is and we'll leave it as is until someone may have the time to completely revise this.

balloob replied:
Having this option in the config flow is weird and should be removed. I would instead always add the camera entity but have it be disabled by default. The only sensor that should be enabled by default is the weather one.

frenck replied:
Just got some input from @MartinHjelmare on Discord as well. Remove the camera from the config flow, create a camera entity, and disable it by default. Put the camera country selector in the options flow so one can change it.

RobBie1221 replied:
Thanks for the input, I can work with that.

RobBie1221 commented:
Thanks for the comments, I'll have a look and address them.

RobBie1221 commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_NAME, default=HOME_LOCATION_NAME): str,
+                vol.Required(CONF_DIMENSION, default=DEFAULT_DIMENSION): vol.All(
+                    vol.Coerce(int), vol.Range(min=120, max=700)
+                ),
```
I did not checkin en.json so you are missing the descriptions.

The parameter is used to get the image in this dimension from the endpoint.

RobBie1221 replied:
We are probably better off moving the dimension to an option flow, if needed.

frenck replied:
I agree, remove the size option; just pick the HQ one, 700px is reasonable.

RobBie1221 commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,167 @@
+
+        data_schema = vol.Schema(
+            {
+                vol.Required(CONF_NAME, default=HOME_LOCATION_NAME): str,
```
The name is used on a few places, mainly to set entity names. I think this is a bit legacy, from before unique ids were used (where it's obviously more important to have good entity names). I think we can do without a name, as users can change entity names.

RobBie1221 committed:
Add translation

RobBie1221 committed:
Disable sensors by default

RobBie1221 committed:
Remove integration name from translations

RobBie1221 committed:
Remove import method

RobBie1221 committed:
Drop  selection between platforms, disable camera by default

RobBie1221 committed:
Minor fix in configured_instances

RobBie1221 committed:
Bugfix in weather

RobBie1221 committed:
Rework import

RobBie1221 committed:
Change unique ids of camera

RobBie1221 committed:
Fix in import

RobBie1221 committed:
Fix camera tests

RobBie1221 committed:
Fix sensor test

RobBie1221 committed:
Fix sensor test 2

RobBie1221 committed:
Fix config flow tests

RobBie1221 committed:
Add option flow

RobBie1221 committed:
Add tests for option flow

RobBie1221 committed:
Add import tests

RobBie1221 committed:
Some cleanups

RobBie1221 commented:
I've made the following changes:
- All platforms are now setup. Only weather is enabled, others are disabled by default
- Removed the name key as option
- Added three remaining useful parameters in an option flow (country and delta for camera, timeframe for precipation sensor)
- Removed dimension as option for camera, always use 700

This is ready for review.

ties commented:
If the comments were addressed (which it looks like), perhaps @frenck can take a look?

frenck commented on tests/components/buienradar/test_camera.py:
```diff
@@ -1,33 +1,54 @@
 
-def radar_map_url(dim: int = 512, country_code: str = "NL") -> str:
+def radar_map_url(country_code: str = "NL") -> str:
     """Build map url, defaulting to 512 wide (as in component)."""
```
```suggestion
    """Build map URL."""
```

frenck commented on tests/components/buienradar/test_camera.py:
```diff
@@ -1,33 +1,54 @@
+def radar_map_url(country_code: str = "NL") -> str:
     """Build map url, defaulting to 512 wide (as in component)."""
-    return f"https://api.buienradar.nl/image/1.0/RadarMap{country_code}?w={dim}&h={dim}"
+    return f"https://api.buienradar.nl/image/1.0/RadarMap{country_code}?w={700}&h={700}"
```
```suggestion
    return f"https://api.buienradar.nl/image/1.0/RadarMap{country_code}?w=700&h=700"
```

frenck commented on homeassistant/components/buienradar/__init__.py:
```diff
@@ -1 +1,162 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up buienradar from a config entry."""
+    hass.data[DOMAIN][entry.entry_id] = {}
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    listener = entry.add_update_listener(async_update_options)
+    hass.data[DOMAIN][entry.entry_id][DATA_LISTENER] = listener
```
After rebase, can be replaced by:

```suggestion
    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
    listener = entry.add_update_listener(async_update_options)
    hass.data[DOMAIN][entry.entry_id] = {DATA_LISTENER: listener}
```

frenck commented on homeassistant/components/buienradar/__init__.py:
```diff
@@ -1 +1,162 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    listener = hass.data[DOMAIN][entry.entry_id][DATA_LISTENER]
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+
+    if unload_ok:
+        listener()
+
+        hass.data[DOMAIN].pop(entry.entry_id)
```
After rebase, can be replaced by:

```suggestion
	unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        hass.data[DOMAIN][entry.entry_id][DATA_LISTENER]()
        hass.data[DOMAIN].pop(entry.entry_id)
```

frenck commented on homeassistant/components/buienradar/config_flow.py:
```diff
@@ -0,0 +1,131 @@
+    """Handle a config flow for buienradar."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
```
This is no longer used.

```suggestion
```

frenck reviewed (commented):
Tested it, looks good!

It does need a rebase and a couple of small suggestions made after rebase has been done to simplify the code just a tiny bit more.

After that, I think it should be ready to go.

RobBie1221 commented:
Thanks, I'll have a look at it

frenck commented:
@RobBie1221 I have a rebased branch (from testing), want me to push it up into this PR?

frenck commented:
Done 👍

RobBie1221 committed:
Apply suggestions from code review

Apply code suggestions

Co-authored-by: Franck Nijhof <git@frenck.dev>

RobBie1221 committed:
Fix isort,black,mypy

RobBie1221 commented:
Changes are merged, CI is happy again. Docs PR is  also outdated, if we feel this is close to ready I'll put in some effort to rewrite it to current implementation.

frenck committed:
Small tweaks and added typing to new parts

frenck commented:
Added some small tweaks to new/changed parts (complete typing) and some small simplifications.
It isn't fully typed by far, but at least the new parts added are now complete.

frenck reviewed (approved):
Tested again, still works great. Thanks, @RobBie1221 👍 

Going to ask for a second look/opinion from a fellow reviewer, but I think it's great 🎉

MartinHjelmare commented on tests/components/buienradar/test_camera.py:
```diff
@@ -1,34 +1,55 @@
-    return f"https://api.buienradar.nl/image/1.0/RadarMap{country_code}?w={dim}&h={dim}"
+async def _setup_config_entry(hass, entry_id):
+    with patch(
+        "homeassistant.components.buienradar.camera.BuienradarCam.entity_registry_enabled_default"
```
Preset the entity registry entry data for this entity instead.

MartinHjelmare commented on tests/components/buienradar/test_config_flow.py:
```diff
@@ -0,0 +1,130 @@
+    )
+    entry.add_to_hass(hass)
+
+    result = await hass.config_entries.options.async_init(entry.entry_id)
```
Set up the integration by setting up the config entry before starting the options flow. It's important to make sure the config flow module is loaded properly. Use `hass.config_entries.async_setup`.

MartinHjelmare commented on tests/components/buienradar/test_sensor.py:
```diff
@@ -1,26 +1,32 @@
+        state = hass.states.get(f"sensor.buienradar_{cond}")
         assert state.state == "unknown"
+
+    await hass.config_entries.async_unload(mock_entry.entry_id)
```
Why do we unload the config entry here?

RobBie1221 commented on homeassistant/components/buienradar/weather.py:
```diff
@@ -35,59 +34,51 @@
+    ATTR_CONDITION_CLOUDY: ("c", "p"),
+    ATTR_CONDITION_FOG: ("d", "n"),
+    ATTR_CONDITION_HAIL: (),
+    ATTR_CONDITION_LIGHTNING: ("g"),
```
I tried using arrays, but that made mypy unhappy: https://github.com/python/mypy/issues/8706

I didn't really understand the annotation suggestion, so I changed to tuples.

RobBie1221 commented on homeassistant/components/buienradar/camera.py:
```diff
@@ -192,3 +187,8 @@ async def async_camera_image(self) -> bytes | None:
+
+    @property
+    def entity_registry_enabled_default(self) -> bool:
+        """Return if the entity should be enabled when first added to the entity registry."""
```
I'll change it. Just an FYI, this exact line I copied from another integration and was pasted apparantly in 30 or so other integrations.

RobBie1221 replied:
Correction, I copied it from entity.py

RobBie1221 committed:
Fix review comments (1)

RobBie1221 committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

RobBie1221 committed:
Fix review comments (2)

RobBie1221 committed:
Merge branch 'buienradar_cfg_flow' of https://github.com/RobBie1221/core into buienradar_cfg_flow

RobBie1221 committed:
Fix issues

RobBie1221 committed:
Fix unique id

RobBie1221 committed:
Improve tests

RobBie1221 committed:
Extend tests

RobBie1221 commented:
I think I got them all :-)

RobBie1221 committed:
Fix issue with unload

MartinHjelmare commented on tests/components/buienradar/test_camera.py:
```diff
@@ -12,29 +11,38 @@
     CONF_LONGITUDE,
     HTTP_INTERNAL_SERVER_ERROR,
 )
+from homeassistant.helpers.entity_registry import async_get_registry
```
Please use `async_get` to get the entity registry instead.

MartinHjelmare commented on homeassistant/components/buienradar/__init__.py:
```diff
@@ -48,17 +42,13 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     return True
 
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```
Please test the setup and unload of the integration in the init module tests.

MartinHjelmare commented on homeassistant/components/buienradar/sensor.py:
```diff
@@ -186,36 +185,24 @@
 
-CONF_TIMEFRAME = "timeframe"
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Please keep the platform schema. We normally log a warning in `async_setup_platform` that the configuration.yaml is deprecated.

MartinHjelmare commented on homeassistant/components/buienradar/sensor.py:
```diff
@@ -225,12 +212,14 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
+    async_add_entities(entities)
 
-    data = BrData(hass, coordinates, timeframe, dev)
+    data = BrData(hass, coordinates, timeframe, entities)
```
A future improvement would be to either use the data update coordinator or the dispatch helper to signal the entities to update instead of accessing the entities from outside. The latter is not a good pattern.

RobBie1221 replied:
I agree, there is more room for improvement.

MartinHjelmare commented on homeassistant/components/buienradar/weather.py:
```diff
@@ -35,59 +34,56 @@
+    ATTR_CONDITION_EXCEPTIONAL: (),
 }
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
```
Please keep the platform schema.

RobBie1221 committed:
Address review comments

RobBie1221 committed:
Add warning when loading platform

RobBie1221 committed:
Add load/unload test

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare reviewed (commented):
Sorry, I noticed some things that we should address in the tests. It seems there's no I/O directly during config entry setup but the platforms will eventually do I/O so it's safest to patch that regardless.

RobBie1221 commented:
I'll address these in a follow-up PR

RobBie1221 commented on tests/components/buienradar/test_camera.py:
```diff
@@ -1,34 +1,63 @@
 
 async def test_fetching_url_and_caching(aioclient_mock, hass, hass_client):
     """Test that it fetches the given url."""
     aioclient_mock.get(radar_map_url(), text="hello world")
```
@MartinHjelmare Thinking about your review comments, in this test we patch the radar url, but not the sensor/weather url. Wouldn't that be wrong as well? It's setting up the ConfigEntry, so also weather and sensor.

MartinHjelmare replied:
The aioclient_mock will patch the session so there will be no real requests going out.
